dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp

```diff
@@ -0,0 +1,462 @@
+{
+    DWORD dwRequired;
+
+    TRACE("(%p, %u, %p, %p)\n", this, BufferSize, Buffer, RequiredSize);
+
+    dwRequired = (DWORD)(wcslen(m_deletedFile.FileNameW) + 1) * sizeof(WCHAR);
+    if (RequiredSize)
+        *RequiredSize = dwRequired;
+
+    if (BufferSize == 0 && !Buffer)
+        return S_OK;
+
+    if (BufferSize < dwRequired)
+        return E_OUTOFMEMORY;
+    CopyMemory(Buffer, m_deletedFile.FileNameW, dwRequired);
+    return S_OK;
+}
+
+STDMETHODIMP RecycleBin5File::GetTypeName(SIZE_T BufferSize, LPWSTR Buffer, SIZE_T *RequiredSize)
+{
+    HRESULT hr;
+    DWORD dwRequired;
+    DWORD dwAttributes;
+    SHFILEINFOW shFileInfo;
+
+    TRACE("(%p, %u, %p, %p)\n", this, BufferSize, Buffer, RequiredSize);
+
+    hr = GetAttributes(&dwAttributes);
+    if (!SUCCEEDED(hr))
+        return hr;
+
+    hr = SHGetFileInfoW(m_FullName, dwAttributes, &shFileInfo, sizeof(shFileInfo), SHGFI_TYPENAME | SHGFI_USEFILEATTRIBUTES);
+    if (!SUCCEEDED(hr))
+        return hr;
+
+    dwRequired = (DWORD)(wcslen(shFileInfo.szTypeName) + 1) * sizeof(WCHAR);
+    if (RequiredSize)
+        *RequiredSize = dwRequired;
+
+    if (BufferSize == 0 && !Buffer)
+        return S_OK;
+
+    if (BufferSize < dwRequired)
+        return E_OUTOFMEMORY;
+    CopyMemory(Buffer, shFileInfo.szTypeName, dwRequired);
+    return S_OK;
+}
+
+STDMETHODIMP RecycleBin5File::Delete()
+{
+    TRACE("(%p)\n", this);
+    return m_recycleBin->Delete(m_FullName, &m_deletedFile);
+}
+
+STDMETHODIMP RecycleBin5File::Restore()
+{
+    TRACE("(%p)\n", this);
+    return m_recycleBin->Restore(m_FullName, &m_deletedFile);
+}
+
+RecycleBin5File::RecycleBin5File(
+    IN IRecycleBin5 *prb,
+    IN LPCWSTR Folder,
+    IN PDELETED_FILE_RECORD pDeletedFile,
+    IN OUT LPWSTR pszFullName)
+    : m_ref(1)
+    , m_recycleBin(prb)
+    , m_deletedFile(*pDeletedFile)
+    , m_FullName(pszFullName)
+{
+    m_recycleBin->AddRef();
+}
+
+static HRESULT
+RecycleBin5File_Constructor(
+    IN IRecycleBin5 *prb,
+    IN LPCWSTR Folder,
+    IN PDELETED_FILE_RECORD pDeletedFile,
+    OUT IRecycleBinFile **ppFile)
+{
+    if (!ppFile)
+        return E_POINTER;
+
+    LPCWSTR Extension = wcsrchr(pDeletedFile->FileNameW, '.');
+    WCHAR szFullName[MAX_PATH];
+    wsprintfW(szFullName, L"%s\\D%c%lu%s", Folder, pDeletedFile->dwDriveNumber + 'a', pDeletedFile->dwRecordUniqueId, Extension);
+    if (GetFileAttributesW(szFullName) == INVALID_FILE_ATTRIBUTES)
+        return E_FAIL;
+
+    LPWSTR pszFullName;
+    HRESULT hr = SHStrDup(szFullName, &pszFullName);
+    if (FAILED(hr))
+        return hr;
+
+    RecycleBin5File *pThis = new RecycleBin5File(prb, Folder, pDeletedFile, pszFullName);
+    if (!pThis)
+        return E_OUTOFMEMORY;
+
+    *ppFile = static_cast<IRecycleBinFile *>(pThis);
+    return S_OK;
+}
+
+class RecycleBin5Enum : public IRecycleBinEnumList
+{
+public:
+    RecycleBin5Enum(
+        IN IRecycleBin5 *prb,
+        IN HANDLE hInfo,
+        IN OUT INFO2_HEADER *pInfo,
```

Comment:
All the `INFO2_HEADER *` --> `PINFO2_HEADER`

Reply:
Fixed.